subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java

```diff
@@ -56,9 +56,20 @@ public FileNameDerivingClassNameConverter(SourceFileClassNameConverter delegate,
             return sourcePaths;
         }
 
+        Set<String> paths = fileExtensions.stream()
+            .map(fileExtension -> className.replace('.', '/') + fileExtension)
+            .collect(Collectors.toSet());
+
+        // Classes with $ may be inner classes
         int innerClassIdx = className.indexOf("$");
-        String baseName = innerClassIdx > 0 ? className.substring(0, innerClassIdx) : className;
-        return fileExtensions.stream().map(fileExtension -> baseName.replace('.', '/') + fileExtension).collect(Collectors.toSet());
+        if (innerClassIdx > 0) {
+            String baseName = className.substring(0, innerClassIdx);
+            paths.addAll(fileExtensions.stream()
+                .map(fileExtension -> baseName.replace('.', '/') + fileExtension)
+                .collect(Collectors.toSet()));
+        }
```

Comment:
```suggestion
            fileExtensions.stream()
                .map(fileExtension -> baseName.replace('.', '/') + fileExtension)
                .forEach(paths::add)
        }
```

Nit: this would avoid creating a second hash set.

Reply:
nice, used that